### PR TITLE
[Add] NanoVLM Blog Post

### PR DESCRIPTION
**Pull request by ariG23498**
@andimarafioti @merveenoyan @sergiopaniego @pcuenca @Vaibhavs10 @lusxvr

### CODE REVIEW

File: nanovlm.md [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+print("\nInput:\n ", args.prompt, "\n\nOutputs:")
+for i in range(args.generations):
+    gen = model.generate(tokens, img_t, max_new_tokens=args.max_new_tokens)
+    out = tokenizer.batch_decode(gen, skip_special_tokens=True)[0]
+    print(f"  >> Generation {i+1}: {out}")
+```
+
+We create the model and set it to `eval`. Initialize the tokenizer, which tokenizes the text prompt,
+and the image processor, which  is used to process the images. The next step is to process the inputs
+and run `model.generate` to generate the output text. Finally, decode the output using  `batch_decode`.
```
Comment (andimarafioti) [inline nanovlm.md:302]: Should we show the results and the image we use? Like show the cat picture and some generations? Maybe we could have a couple of pictures and generations

File: nanovlm.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+By keeping the codebase lightweight and readable, nanoVLM aims to serve as both a learning tool and a foundation you can build upon. Whether you’re looking to understand how multi-modal inputs are aligned, or you want to train a VLM on your own dataset, this repository gives you a head start.
+
+If you try it out, build on top of it, or just have questions we’d love to hear from you. Happy tinkering!
+
+## References
+
+1. [GitHub - huggingface/nanoVLM: The simplest, fastest repository for training/finetuning small-sized VLMs.](https://github.com/huggingface/nanoVLM)  
+2. [A Dive into Vision-Language Models](https://huggingface.co/blog/vision_language_pretraining)  
+3. [Vision Language Models Explained](https://huggingface.co/blog/vlms)
```
Comment (andimarafioti) [inline nanovlm.md:319]: Maybe add the SmolVLm paper here?

**Review by andimarafioti — COMMENTED** [review]
Great work! I left some comments :)

File: nanovlm.md [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+## What is a Vision Language Model?
+
+As the name suggests, a Vision Language Model (VLM) is a multi-modal model that processes two
+modalities: vision and text.These models typically take images and/or text as input and generate text as output.
+
+Generating text (output) conditioned on the understanding of images and texts (inputs) is a powerful paradigm.
+It enables a wide range of applications, from image captioning and object detection to answering
+questions about visual content (as shown in the table below). One thing to note is that nanoVLM
+focuses only on Visual Question Answering as the training objective.
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:64]: I agree with your PoV, but I had a lot of feedback on this particular para being confusing, hence I wanted to be explicit.

File: nanovlm.md [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+| Image (input) | Query (Text as input) | Generation | Task |
+| :---- | :---- | :---- | :---- |
+| ![an image of a cat](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg) | Caption the image | Two cats lying down on a bed with remotes near them | Captioning |
+|  | Detect the objects in the image | `<locxx><locxx><locxx><locxx>` | Object Detection |
+|  | Segment the objects in the image | `<segxx><segxx><segxx>` | Semantic Segmentation |
+|  | How many cats are in the image? | 2 | Visual Question Answering |
+
+> [!TIP]  
+> If you are interested in learning more about VLMs, we strongly recommend reading our latest
+blog on the topic: [Vision Language Models (Better, Faster, Stronger)](https://huggingface.co/blog/vlms-2025)
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:75]: Great catch!

File: nanovlm.md [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+We model nanoVLM after two well known and widely used architectures. Our vision backbone
+(`models/vision_transformer.py`) is the standard vision transformer, more specifically Google’s
+[SigLIP](https://huggingface.co/docs/transformers/en/model_doc/siglip) vision encoder. Our language
+backbone follows the [Llama 3](https://huggingface.co/docs/transformers/en/model_doc/llama3) architecture.
+
+The vision and text modalities are *aligned* using a Modality Projection module. This module takes the
+image embeddings produced by the vision backbone as input, and transforms them into embeddings
+compatible with the text embeddings from the embedding layer of the language model. These embeddings
+are then concatenated and fed into the language decoder. The Modality Projection module consists of a
+pixel shuffle operation followed by a linear layer.
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:124]: All thanks to @sergiopaniego!

File: nanovlm.md [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+| ![diagram of the model architecture](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/architecture.png) |
+| :--: |
+| The architecture of the model (Source: Authors) |
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:128]: For sure. I will add the paper in the references section.

File: nanovlm.md [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+At the heart of the data pipeline is the `get_dataloaders` function. It:
+
+* Loads datasets via Hugging Face’s `load_dataset` API.  
+* Combines and shuffles multiple datasets (if provided).  
+* Applies a train/val split via indexing.  
+* Wraps them in custom datasets (`VQADataset`, `MMStarDataset`) and collators (`VQACollator`, `MMStarCollator`).
+
+> [!TIP]  
+> A helpful flag here is `data_cutoff_idx`, useful for debugging on small subsets.
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:190]: Would you like me to remove this? 😆

File: nanovlm.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+You can easily push them using:
+
+```python
+model.push_to_hub("hub/id")
+```
+
+## Run inference on a pre-trained model
+
+Using nanoVLM as the toolkit, we have trained a [model and published it to Hub](https://huggingface.co/lusxvr/nanoVLM-222M).
+We have used the `SigLIP-B/16-224-85M` and `HuggingFaceTB/SmolLM2-135M` as backbones. The model was
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:249]: I got this from the README of nanoVLM. I will use the checkpoint shared by you then.

File: nanovlm.md [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+Inspired by [Andrej Karpathy](https://karpathy.ai/)’s [nanoGPT](https://github.com/karpathy/nanoGPT),
+Hugging Face offers a similar project for the vision domain: [**nanoVLM**](https://github.com/huggingface/nanoVLM).
+This repository is designed to be the simplest way to get started with training your own
+Vision Language Model (VLM) using pure PyTorch.
+
+At its heart, nanoVLM is a **toolkit** that helps you build and train a model that can understand both
+images and text, and then generate text based on that. The beauty of nanoVLM lies in its *simplicity*.
+The entire codebase is intentionally kept *minimal* and *readable*, making it perfect for beginners or
+anyone who wants to peek under the hood of VLMs without getting overwhelmed.
+
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply nanovlm.md:27]: I feel like this should make it a little redundant. I have added the note on being lightweight in the previous para.

Could you review the changes and see how that works? If you still want me to rephrase the current para, I would happily do it.

File: nanovlm.md [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+modalities: vision and text. These models typically take images and/or text as input and generate text as output.
+
+Generating text (output) conditioned on the understanding of images and texts (inputs) is a powerful paradigm.
+It enables a wide range of applications, from image captioning and object detection to answering
+questions about visual content (as shown in the table below). One thing to note is that nanoVLM
+focuses only on Visual Question Answering as the training objective.
+
+| Image (input) | Query (Text as input) | Generation | Task |
+| :---- | :---- | :---- | :---- |
+| ![an image of a cat](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg) | Caption the image | Two cats lying down on a bed with remotes near them | Captioning |
```
Comment (lusxvr) [inline nanovlm.md:70]: Is it possible to have the image be over all the lines of the table? I.e. merge the first column into one row?

File: nanovlm.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (andimarafioti) [reply nanovlm.md:190]: no no, it's great, I wasn't so hands on with nanovlm and really didn't know what this was for

File: nanovlm.md [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+|  | Detect the objects in the image | `<locxx><locxx><locxx><locxx>` | Object Detection |
+|  | Segment the objects in the image | `<segxx><segxx><segxx>` | Semantic Segmentation |
+|  | How many cats are in the image? | 2 | Visual Question Answering |
```
Comment (sergiopaniego) [inline nanovlm.md:73]:
```suggestion
<table>
  <tr>
    <td rowspan="4"><img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/cat.jpg" alt="an image of a cat" width="200"/></td>
    <td>Caption the image</td>
    <td>Two cats lying down on a bed with remotes near them</td>
    <td>Captioning</td>
  </tr>
  <tr>
    <td>Detect the objects in the image</td>
    <td><code>&lt;locxx&gt;&lt;locxx&gt;&lt;locxx&gt;&lt;locxx&gt;</code></td>
    <td>Object Detection</td>
  </tr>
  <tr>
    <td>Segment the objects in the image</td>
    <td><code>&lt;segxx&gt;&lt;segxx&gt;&lt;segxx&gt;</code></td>
    <td>Semantic Segmentation</td>
  </tr>
  <tr>
    <td>How many cats are in the image?</td>
    <td>2</td>
    <td>Visual Question Answering</td>
  </tr>
</table>
```

File: nanovlm.md [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+image embeddings produced by the vision backbone as input, and transforms them into embeddings
+compatible with the text embeddings from the embedding layer of the language model. These embeddings
+are then concatenated and fed into the language decoder. The Modality Projection module consists of a
+pixel shuffle operation followed by a linear layer.
+
+| ![diagram of the model architecture](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/nanovlm/architecture.png) |
+| :--: |
+| The architecture of the model (Source: Authors) |
+
+[Pixel shuffle](https://arxiv.org/abs/1609.05158) reduces the number of image tokens, which helps
```
Comment (sergiopaniego) [inline nanovlm.md:130]:
```suggestion
[Pixel shuffle](https://huggingface.co/papers/1609.05158) reduces the number of image tokens, which helps
```

File: nanovlm.md [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+
+If you try it out, build on top of it, or just have questions we’d love to hear from you. Happy tinkering!
+
+## References
+
+1. [GitHub - huggingface/nanoVLM: The simplest, fastest repository for training/finetuning small-sized VLMs.](https://github.com/huggingface/nanoVLM)  
+2. [A Dive into Vision-Language Models](https://huggingface.co/blog/vision_language_pretraining)  
+3. [Vision Language Models Explained](https://huggingface.co/blog/vlms)  
+4. [Vision Language Models (Better, faster, stronger)](https://huggingface.co/blog/vlms-2025)
+5. [SmolVLM: Redefining small and efficient multimodal models](https://arxiv.org/abs/2504.05299)
```
Comment (sergiopaniego) [inline nanovlm.md:326]:
```suggestion
5. [SmolVLM: Redefining small and efficient multimodal models](https://huggingface.co/papers/2504.05299)
```

File: nanovlm.md [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+In this blog post, we walked through what VLMs are, explored the architecture choices that power nanoVLM, and unpacked the training and inference workflows in detail.
+
+By keeping the codebase lightweight and readable, nanoVLM aims to serve as both a learning tool and a foundation you can build upon. Whether you’re looking to understand how multi-modal inputs are aligned, or you want to train a VLM on your own dataset, this repository gives you a head start.
+
+If you try it out, build on top of it, or just have questions we’d love to hear from you. Happy tinkering!
+
+## References
+
+1. [GitHub - huggingface/nanoVLM: The simplest, fastest repository for training/finetuning small-sized VLMs.](https://github.com/huggingface/nanoVLM)  
+2. [A Dive into Vision-Language Models](https://huggingface.co/blog/vision_language_pretraining)
```
Comment (sergiopaniego) [inline nanovlm.md:323]: I'd reorder the VLM blogs by release date.

**Review by sergiopaniego — APPROVED** [review]
🚀

File: nanovlm.md (hunk 10 above)
Comment (sergiopaniego) [reply nanovlm.md:70]: Yes (using HTML). I've added it on a suggestion

File: nanovlm.md [hunk 15]
```diff
@@ -13,10 +13,12 @@ authors:
...
-Inspired by [Andrej Karpathy](https://karpathy.ai/)’s [nanoGPT](https://github.com/karpathy/nanoGPT),
-Hugging Face offers a similar project for the vision domain: [**nanoVLM**](https://github.com/huggingface/nanoVLM).
-This repository is designed to be the simplest way to get started with training your own
-Vision Language Model (VLM) using pure PyTorch.
+[**nanoVLM**](https://github.com/huggingface/nanoVLM) is the *simplest* way to get started with
+**training** your very own Vision Language Model (VLM) using pure PyTorch. It is lightweight *toolkit*
+which allows you to launch a VLM training on a [free tier colab notebook](https://colab.research.google.com/github/huggingface/nanoVLM/blob/main/nanoVLM.ipynb).
+
+> We were inspired by [Andrej Karpathy](https://karpathy.ai/)’s [nanoGPT](https://github.com/karpathy/nanoGPT),
+and provide a similar project for the vision domain.
```
[the remark this answers is not in the document]
Comment (lusxvr) [reply nanovlm.md:21]: image

**Review by lusxvr — APPROVED** [review]
Looking great!

File: nanovlm.md [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+
+If `log_wandb` is enabled, training stats like `batch_loss`,  `val_loss`, `accuracy`, and `tokens_per_second`
+are logged to Weights & Biases for real-time tracking.
+
+Runs are auto-named using metadata like sample size, batch size, epoch count, learning rates, and the date,
+all handled by the helper `get_run_name`.
+
+**Push to Hub**
+
+The script doesn’t automatically push to the Hugging Face Hub, but since models are saved via:
```
Comment (merveenoyan) [inline nanovlm.md:234]: I really think we should make this mandatory, previously @Vaibhavs10 commented the same

File: nanovlm.md [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+> A helpful flag here is `data_cutoff_idx`, useful for debugging on small subsets.
+
+**Model Initialization**
+
+The model is built via the `VisionLanguageModel` class. If you're resuming from a checkpoint, it’s as easy as:
+
+```python
+from models.vision_language_model import VisionLanguageModel
+
+model = VisionLanguageModel.from_pretrained(path)
```
Comment (merveenoyan) [inline nanovlm.md:199]:
```suggestion
model = VisionLanguageModel.from_pretrained(model_path)
```

File: nanovlm.md [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+
+Runs are auto-named using metadata like sample size, batch size, epoch count, learning rates, and the date,
+all handled by the helper `get_run_name`.
+
+**Push to Hub**
+
+The script doesn’t automatically push to the Hugging Face Hub, but since models are saved via:
+
+```python
+model.save_pretrained(path)
```
Comment (merveenoyan) [inline nanovlm.md:237]:
```suggestion
model.save_pretrained(save_path)
```

**Review by merveenoyan — APPROVED** [review]
I left very general comments, looking great anyway!

File: nanovlm.md (hunk 17 above)
Comment (merveenoyan) [reply nanovlm.md:199]: nit: ckpt load and the new model save path are same otherwise

File: nanovlm.md (hunk 16 above)
Comment (lusxvr) [reply nanovlm.md]: I agree with the point that the users should push to the hub automatically, the problem is more that we also use the same code quite extensively ourselves so if we make it the default to push to the hub, every experiment we run gets pushed there too, or we would have to change this for every new commit we do on our machines vs in the repo.

I thought about a compromise that we could change the colab notebook (since I think a lot of people start with this) to push automatically to the hub (since we do not use it), and then leave it as an option in the normal codebase. 

What do you think?

Comment (ariG23498) [reply nanovlm.md]: Starting with colab is a very nice decision.

File: nanovlm.md [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+- [Conclusion](#conclusion)
+- [References](#references)
+
+## TL;DR
+
+You can start training a Vision Language Model using our nanoVLM toolkit by following these steps:
+
+```bash
+# Clone the repo
+git clone https://github.com/huggingface/transformers.git
```
Comment (kashif) [inline nanovlm.md:48]:
```suggestion
git clone https://github.com/huggingface/nanoVLM.git
```

File: nanovlm.md (hunk 16 above)
Comment (andimarafioti) [reply nanovlm.md]: We can definitely change it in our configs if we don't want this to happen, it can just be the default, but have an option to deactivate it.
